besu/src/main/java/org/hyperledger/besu/controller/QbftBesuControllerBuilder.java

```diff
@@ -0,0 +1,267 @@
+import org.hyperledger.besu.ethereum.core.Util;
+import org.hyperledger.besu.ethereum.eth.EthProtocol;
+import org.hyperledger.besu.ethereum.eth.manager.EthProtocolManager;
+import org.hyperledger.besu.ethereum.eth.sync.state.SyncState;
+import org.hyperledger.besu.ethereum.eth.transactions.TransactionPool;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.ethereum.p2p.config.SubProtocolConfiguration;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+import org.hyperledger.besu.util.Subscribers;
+
+import java.util.Collections;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class QbftBesuControllerBuilder extends BesuControllerBuilder {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private BftEventQueue bftEventQueue;
+  private BftConfigOptions bftConfig;
+  private ValidatorPeers peers;
+  private final BlockInterface blockInterface = new BftBlockInterface();
+
+  @Override
+  protected void prepForBuild() {
+    bftConfig = genesisConfig.getConfigOptions(genesisConfigOverrides).getBftConfigOptions();
+    bftEventQueue = new BftEventQueue(bftConfig.getMessageQueueLimit());
+  }
+
+  @Override
+  protected JsonRpcMethods createAdditionalJsonRpcMethodFactory(
+      final ProtocolContext protocolContext) {
+    return new IbftJsonRpcMethods(protocolContext);
+  }
+
+  @Override
+  protected SubProtocolConfiguration createSubProtocolConfiguration(
+      final EthProtocolManager ethProtocolManager) {
+    return new SubProtocolConfiguration()
+        .withSubProtocol(EthProtocol.get(), ethProtocolManager)
+        .withSubProtocol(
+            QbftSubProtocol.get(),
+            new BftProtocolManager(
+                bftEventQueue, peers, QbftSubProtocol.QBFV1, QbftSubProtocol.get().getName()));
+  }
+
+  @Override
+  protected MiningCoordinator createMiningCoordinator(
+      final ProtocolSchedule protocolSchedule,
+      final ProtocolContext protocolContext,
+      final TransactionPool transactionPool,
+      final MiningParameters miningParameters,
+      final SyncState syncState,
+      final EthProtocolManager ethProtocolManager) {
+    final MutableBlockchain blockchain = protocolContext.getBlockchain();
+    final BftExecutors bftExecutors = BftExecutors.create(metricsSystem);
+
+    final Address localAddress = Util.publicKeyToAddress(nodeKey.getPublicKey());
+    final BftBlockCreatorFactory blockCreatorFactory =
+        new BftBlockCreatorFactory(
+            gasLimitCalculator,
+            transactionPool.getPendingTransactions(),
+            protocolContext,
+            protocolSchedule,
+            miningParameters,
+            localAddress,
+            bftConfig.getMiningBeneficiary().map(Address::fromHexString).orElse(localAddress));
+
+    // NOTE: peers should not be used for accessing the network as it does not enforce the
```

Comment:
can this comment be removed? I mean, its sensible, but its not useful right here.

Reply:
maybe more useful on line 139/140?

Reply:
done